### PR TITLE
Capture stdout/stderr of spawned components

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Captures the standard out and standard error of spawned components of the command runtime. Parses their output and sends them through the logger at the underlying level bypassing any log level setting for the Elastic Agent. This allows the each component to have a lower log level than the Elastic Agent itself (useful for debugging).

This at the moment does not handle logging from the any service runtime (aka. Endpoint Security). I will look into that in a later branch.

## Why is it important?

Provides a unified view of what is happening with the components that the Elastic Agent is running. Before that was hidden in files now a unified file provides all logs from the elastic-agent and the spawned components.

**Note**: Without this change no logs from the running components in 8.6 get sent to Fleet with monitoring turned on.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Run the elastic-agent locally with `elastic-agent run -e` and you will see log messages from the spawned components shown in the output.

## Related issues

- Relates #221

### CODE REVIEW

**Comment by blakerouse** [comment]
This is finally working how I wanted it to work. Took longer than I would have liked but ran into many different issues as well as just many different times refactoring it until I was actually happy with the implementation.

This has a requirement of https://github.com/elastic/beats/pull/33800 being merged first, before this one can be merged. That processor is required for this to work properly.

File: internal/pkg/agent/application/monitoring/v1_monitor.go
Comment (cmacknz) [inline internal/pkg/agent/application/monitoring/v1_monitor.go:389]: Are the destination datastreams all the same as they were before this change? Are there any new datastreams created here?

I haven't tested this locally yet, I just want to do an early check that we haven't introduced anything that could be considered a breaking change.

Comment (blakerouse) [reply internal/pkg/agent/application/monitoring/v1_monitor.go:389]: The resulting datastreams are different in that now they are grouped by spawned component. This is done on purpose and a design decision of V2. The allows logs from each component that now has its own status to be grouped properly. So you can correlate status of a component to the exact logs of that component.

File: specs/auditbeat.spec.yml
Comment (cmacknz) [inline specs/auditbeat.spec.yml:27]: Is the switch from `debug` to `info` here intentional? Looks like it was set to debug on 8.5 as well https://github.com/elastic/elastic-agent/blob/6f08495275dba06a20aae3c8f62de69864893b90/internal/spec/auditbeat.yml#L8

Comment (cmacknz) [reply specs/auditbeat.spec.yml:27]: I see the same for Filebeat and Metricbeat, were we just default to debug logs being enabled all the time?

**Review by cmacknz — COMMENTED** [review]
Can you describe the bug this solves in the current V2 logging implementation so we can merge it to 8.6?

If this isn't merged by EOD today (November 28th) it won't be in 8.6 BC4 scheduled for November 29th. This means our QA team won't be testing it as part of the 8.6 release until the next BC on December 6th.

Edit: I think switching off debug logs by default can be considered a bug fix.

File: specs/auditbeat.spec.yml
Comment (cmacknz) [reply specs/auditbeat.spec.yml:27]: Agreed, this is a good thing. I am more surprised we were defaulting logs to debug, I made this comment before I mentally worked through the consequences of this. There is likely a performance impact to defaulting to debug because of this.

**Comment by blakerouse** [comment]
Added it to the description of the PR, but without this change 8.6 is broken.

Below is the copy of what was placed in the description:

Without this change no logs from the running components in 8.6 get sent to Fleet with monitoring turned on.

File: changelog/fragments/1669236059-Capture-stdout-stderr-of-all-spawned-components-to-simplify-logging.yaml
Comment (cmacknz) [inline changelog/fragments/1669236059-Capture-stdout-stderr-of-all-spawned-components-to-simplify-logging.yaml:14]: I think we should mention that the default log level for Beat sub-processes is not longer debug.

Comment (blakerouse) [reply changelog/fragments/1669236059-Capture-stdout-stderr-of-all-spawned-components-to-simplify-logging.yaml:14]: That is not going to fit in the 80ish character limit. Should I add that as another changelog entry? Or just make it really long?

File: internal/pkg/agent/application/monitoring/v1_monitor.go
Comment (cmacknz) [reply internal/pkg/agent/application/monitoring/v1_monitor.go:389]: 👍

File: internal/pkg/agent/application/monitoring/v1_monitor.go
Comment (michalpristas) [inline internal/pkg/agent/application/monitoring/v1_monitor.go:374]: i'm just wondering, have you checked overall CPU consumption with the move towards stdout/err monitoring? 
seems like std should be a bit better right?

File: internal/pkg/agent/application/monitoring/v1_monitor.go
Comment (michalpristas) [inline internal/pkg/agent/application/monitoring/v1_monitor.go:390]: nice approach

File: pkg/component/runtime/log_writer.go
Comment (michalpristas) [inline pkg/component/runtime/log_writer.go:47]: i'm ok with this, i imagine you considered bytes.SplitN. 
memory consumption would be higher but implementation easier. indexByte also scans whole string for a match so there won't be cpu overhead.

File: internal/pkg/agent/application/monitoring/v1_monitor.go
Comment (blakerouse) [reply internal/pkg/agent/application/monitoring/v1_monitor.go:374]: I don't know about the overall load change, I would say its a net wash. Being that the components no longer need to worry about rotation, syncing files, or worrying about deleting old logs.

I think the benefits of components not needing to really have to worry about logging, rotation, etc; is a win!

File: specs/auditbeat.spec.yml
Comment (michalpristas) [inline specs/auditbeat.spec.yml:27]: yaay

File: pkg/component/runtime/log_writer.go
Comment (michalpristas) [inline pkg/component/runtime/log_writer.go:64]: bytes.TrimSuffix does exactly the same thing

**Review by michalpristas — APPROVED** [review]
overall it looks good to me
let's see if we will run into some edge cases so far it looks it works

File: changelog/fragments/1669236059-Capture-stdout-stderr-of-all-spawned-components-to-simplify-logging.yaml
Comment (blakerouse) [reply changelog/fragments/1669236059-Capture-stdout-stderr-of-all-spawned-components-to-simplify-logging.yaml]: Just got it to fit in the same one.

Comment (cmacknz) [reply changelog/fragments/1669236059-Capture-stdout-stderr-of-all-spawned-components-to-simplify-logging.yaml]: You can use the description field if you need it to be longer than 80 characters.

**Comment by cmacknz** [comment]
Finally got this packaged to test locally. One thing I noticed is that the spawned components contain message fields to easily allow separating them out when analyzing logs locally:

```json
"component":{"binary":"metricbeat","dataset":"elastic_agent.http_metrics_monitoring","id":"http/metrics-monitoring","type":"http/metrics"},"log.origin":{"file.line":721,"file.name":"instance/beat.go"},"service.name":"metricbeat","ecs.version":"1.6.0","ecs.version":"1.6.0"}
```

For example we have `component":{"binary":"metricbeat","dataset":"elastic_agent.http_metrics_monitoring","id":"http/metrics-monitoring","type":"http/metrics"}` and `"service.name":"metricbeat"` to allow looking at the monitoring metricbeat logs only.

We are missing these with logs coming from the agent itself:

```json
{"log.level":"info","@timestamp":"2022-11-28T19:59:12.045Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":645},"message":"Updating running component model","ecs.version":"1.6.0"}
```

I think we would want at least `"service.name":"agent"` to allow filtering for agent specific logs easily e.g. being able to do `rg '"service.name":"agent"'` locally will save a lot of time during SDH analysis.

**Comment by blakerouse** [comment]
@cmacknz We should follow that up in a bug, and not in this PR. That will have to be handled carefully because the base logger is used, so you don't want that field injected into the the other component logs as well, so its not as easy as just adding it to the base initialization.

**Comment by cmacknz** [comment]
Agree that is shouldn't block this PR. I just enrolled my agent in fleet and I'm not seeing any logs for the beats coming through.

My local agent logs are full of "Cannot index event" errors like:

```
{"log.level":"warn","@timestamp":"2022-11-28T20:10:06.398Z","message":"Cannot index event publisher.Event{Content:beat.Event{Timestamp:time.Date(2022, time.November, 28, 20, 9, 50, 574000000, time.UTC), Meta:{\"input_id\":\"filestream-monitoring-agent\",\"raw_index\":\"logs-elastic_agent.filestream_monitoring-default\",\"stream_id\":\"filestream-monitoring-agent\"}, Fields:{\"agent\":{\"ephemeral_id\":\"5a2d7383-1f63-4ae6-8919-caf0273f2514\",\"id\":\"bcce2934-9dc4-4031-89ad-e152984984d0\",\"name\":\"Craigs-MacBook-Pro.local\",\"type\":\"filebeat\",\"version\":\"8.7.0\"},\"component\":{\"binary\":\"filebeat\",\"dataset\":\"elastic_agent.filestream_monitoring\",\"id\":\"filestream-monitoring\",\"type\":\"filestream\"},\"data_stream\":
```

Conveniently it doesn't tell me why it couldn't index the event.

**Comment by cmacknz** [comment]
Looks like this is caused by 403 errors. After looking harder.

`"action [indices:admin/auto_create] is unauthorized for API key id [T4Lav4QBuKu_zclXw8LG] of user [elastic/fleet-server] on indices [logs-elastic_agent.filestream_monitoring-default], this action is granted by the index privileges [auto_configure,create_index,manage,all]`

**Comment by cmacknz** [comment]
Looking at the policy in diagnostics there are no index permissions for `logs-elastic_agent.filestream_monitoring-default`

@blakerouse do we need a change on the fleet side to allow us to write to the new data streams?

**Comment by blakerouse** [comment]
@cmacknz We should have permissions to write to all of `logs-elastic_agent.*`.

**Comment by blakerouse** [comment]
I did do my testing using a standalone agent with superuser password, possible I missed this issue.

**Comment by cmacknz** [comment]
@kpollich are we able to change the index permissions for agent monitoring in an agent policy? We have some new indexes to include as part of V2, but it looks like the current set are just hard coded on the fleet side: https://github.com/elastic/kibana/blob/19413b7daae983b95dbb9f5c7b39cb8f3578ebfa/x-pack/plugins/fleet/server/services/agent_policies/monitoring_permissions.ts#L21-L52

**Comment by kpollich** [comment]
> @kpollich are we able to change the index permissions for agent monitoring in an agent policy? We have some new indexes to include as part of V2, but it looks like the current set are just hard coded on the fleet side: https://github.com/elastic/kibana/blob/19413b7daae983b95dbb9f5c7b39cb8f3578ebfa/x-pack/plugins/fleet/server/services/agent_policies/monitoring_permissions.ts#L21-L52

This should be a pretty small change in Kibana to this hardcoded list:

https://github.com/elastic/kibana/blob/c5f20721e1879f1ebe161b0fa3b207f10ed2b6f7/x-pack/plugins/fleet/common/constants/agent_policy.ts#L15-L28

However, I'd hesitate to make that change during FF for 8.6. Are we just adding a few indices here, or removing old ones? If it's a purely additive change I think that'd be fine, the more I think about it. Happy to push a Kibana PR with an updated list if it's helpful in validating this. 

If it's possible, I'd prefer _not_ to fundamentally alter these permissions during FF, but a purely additive change sounds reasonable to me. Let me know how I can help.

**Comment by cmacknz** [comment]
Tracking the issue with index permissions separately in https://github.com/elastic/elastic-agent/issues/1814

**Pull request merged**
